### PR TITLE
vdk-impala: Classify impala errors as User error if needed

### PR DESCRIPTION
**Pull request by gageorgiev**
This change makes sure that if an error occurs during a job run, the
error is analyzed to check if it was caused by the user. If this is
true, an appropriate error message is printed.

Testing done: unit test and a functional test

Signed-off-by: Gabriel Georgiev <gageorgiev@vmware.com>

### CODE REVIEW

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+def _is_query_memory_error(exception):
+    return errors.exception_matches(
+        exception,
+        classname_with_package="impala.error.OperationalError",
+        exception_message_matcher_regex=".*Memory limit exceeded.*",
+    )
+
+
+def _is_redshift_error(exception):
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py:130]: the redshift user error classification seems out of scope

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+def _is_property_unsupported_value_error(exception):
+    return errors.exception_matches(
+        exception,
+        classname_with_package="ValueError",
+        exception_message_matcher_regex=".*[pP]roperty.*is of unsupported type.*",
+    )
+
+
+def _is_impala_memory_limit_exceeded(exception):
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py:106]: Looking at memory limit exceeded conditions, I am missing 
```
(errors.exception_matches(
        exception,
        classname_with_package="MemoryError",
        exception_message_matcher_regex=".*memory limit.*")
```

Also, could combine with the common `_is_query_memory_error()` one into one memory-related block

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_plugin.py [hunk 3]
```diff
@@ -67,6 +70,18 @@ def initialize_job(self, context: JobContext) -> None:
             lambda: _connection_by_configuration(self._impala_cfg),
         )
 
+    @hookimpl(tryfirst=True)
+    def vdk_exception(self, exception: Exception):
+        if is_impala_error(exception):
+            log_and_throw(
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_plugin.py:76]: One note - looking at the current `vdk_exception()` wiring, any exception raised within such hook implementation, would not be handled by any plugin in the chain but seems to be raised directly instead. Maybe we need to extend the core hookspec docs on exception raising:
```
 @hookspec
 def vdk_exception(self, exception: Exception) -> bool:
        """
        Called in case the CLI is about to exit with error.

        :return: True if exception is handled and does not need to be logged by main
        """
        pass
```

**Review by ivakoleva — COMMENTED** [review]
LGTM, would suggest an additional review and adding the tests

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+from vdk.internal.core import errors
+
+
+def is_impala_error(received_exception):
```
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py:6]:
```suggestion
def is_impala_user_error(received_exception):
```

**Comment by tozka** [comment]
I see it's a draft. But where are the unit tests. 

We also need a functional test - it doesn't need to cover all the cases (if they are covered with unit tests) but it should cover 1 or 2 . The functional test should work similar to this one: https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-plugins/vdk-trino/tests/test_trino_sql.py
It is particularly important to check termination message is correctly set.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_plugin.py [hunk 5]
```diff
@@ -67,6 +70,18 @@ def initialize_job(self, context: JobContext) -> None:
             lambda: _connection_by_configuration(self._impala_cfg),
         )
 
+    @hookimpl(tryfirst=True)
+    def vdk_exception(self, exception: Exception):
```
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_plugin.py:74]: Hooking to vdk_exception may not work since classification happens earlier. This way the termination message will not be udpated correctly. 

See how I've done it for trino: https://github.com/vmware/versatile-data-kit/pull/650/files

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py (hunk 2 above)
Comment (tozka) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_classifier.py:106]:
```
(errors.exception_matches(
        exception,
        classname_with_package="MemoryError",
        exception_message_matcher_regex=".*memory limit.*")
```
This is not an impala error. It is a generic python memory error and it is covered by the core job input error classifier: https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/run/job_input_error_classifier.py#L72

**Comment by ivakoleva** [comment]
> I see it's a draft. But where are the unit tests.
> 
> We also need a functional test - it doesn't need to cover all the cases (if they are covered with unit tests) but it should cover 1 or 2 . The functional test should work similar to this one: https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-plugins/vdk-trino/tests/test_trino_sql.py It is particularly important to check termination message is correctly set.

The functional test and unit tests are now present; the unit tests seem to cover all subcases except `_is_udf_error()`

**Review by ivakoleva — APPROVED** [review]
